apps/web/pages/api/intent-username/index.ts

```diff
@@ -0,0 +1,51 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+import { getSession } from "next-auth/react";
+
+import prisma from "@calcom/prisma";
+import { Prisma } from "@calcom/prisma/client";
+
+import { checkUsername } from "@lib/core/server/checkUsername";
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  const { intentUsername } = req.body;
+  // Check that user is authenticated
+  try {
+    const session = await getSession({ req });
+    const userId = session?.user?.id;
+    const user = await prisma.user.findFirst({ where: { id: userId }, rejectOnNotFound: true });
+    const checkPremiumUsernameResult = await checkUsername(intentUsername);
+
+    if (userId && user) {
+      const userWithMetadata = await prisma.user.findFirst({
+        where: {
+          id: userId,
+        },
+        select: {
+          id: true,
+          metadata: true,
+        },
+      });
+
+      const userMetadata =
+        userWithMetadata && !!userWithMetadata.metadata
+          ? (userWithMetadata.metadata as Prisma.JsonObject)
+          : {};
```

Comment:
We can use zod for this. I think there's a global one on `packages/prisma/zod-utils.ts`